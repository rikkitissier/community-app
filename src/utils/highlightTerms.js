import React, { Component } from "react";
import { Text, View, FlatList, TouchableHighlight, StyleSheet } from "react-native";

/**
 * Fairly na√Æve search highlight parsing.
 * Builds a collection of <Text> components, with matching search terms highlighted. E.g.
 * searching for "test" in "Mattis Parturient Mollis Test Consectetur" should return:
 * <Text>Mattis Parturient Mollis</Text><Text>Test</Text><Text>Consectetur</Text>
 * (though in practice we add styles and spacers)
 *
 * @param 	string 		content 	String containing the content to parse
 * @param 	string 		term 		Search string. Will be split by a space to get individual terms.
 * @return 	array
 */
export default function highlightTerms(content, term, highlightStyle) {
	if (!term || !content) {
		return content;
	}

	const words = content.replace(/[\n\t\r]/g, "").split(" ");
	const terms = term.split(" ").map(term => term.toLowerCase());
	const finalContent = [];
	let currentText = [];
	let counter = 0;

	words.forEach(word => {
		if (terms.indexOf(word.toLowerCase().trim()) !== -1) {
			// If one of our terms is found, put previous text in a <Text>, our term in another <Text>
			if (currentText.length) {
				finalContent.push(<Text key={counter++}>{currentText.join(" ")}</Text>);
				finalContent.push(<Text key={counter++}>&nbsp;</Text>);
				currentText = [];
			}

			finalContent.push(
				<Text key={counter++} style={highlightStyle}>
					{" "}
					{word}{" "}
				</Text>
			);
			finalContent.push(<Text key={counter++}>&nbsp;</Text>);
		} else {
			// Just continue...
			currentText.push(word);
		}
	});

	// If we have any words left, add those
	if (currentText.length) {
		finalContent.push(<Text key={counter++}>{currentText.join(" ")}</Text>);
	}

	return finalContent;
}
