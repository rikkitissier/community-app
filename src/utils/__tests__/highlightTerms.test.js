import React from "react";
import { View, StyleSheet } from "react-native";
import { shallow } from "enzyme";
import _ from "underscore";

import highlightTerms from "../highlightTerms";

const CONTENT = "Cras mattis consectetur purus sit amet fermentum. Cum sociis natoque mattis penatibus et magnis dis parturient montes, nascetur ridiculus mus.";
const HIGHLIGHT_STYLE = { backgroundColor: "red" };

const removeFalsy = (val) => !!val.trim();

describe("highlightTerms utility", () => {
	it("returns correct structure for single VALID search term", () => {	
		const wrapper = shallow(<View>{highlightTerms(CONTENT, "purus", HIGHLIGHT_STYLE)}</View>);
		expect(wrapper).toMatchSnapshot();

		// There should be five text nodes
		expect( wrapper.find("View > Text").length ).toBe(5);
		// The third should be our search term
		expect( _.filter( wrapper.find("View > Text").at(2).prop('children'), removeFalsy )[0] ).toBe("purus");
	});

	it("returns correct structure for multiple VALID search term", () => {	
		const wrapper = shallow(<View>{highlightTerms(CONTENT, "mattis", HIGHLIGHT_STYLE)}</View>);
		expect(wrapper).toMatchSnapshot();

		// There should be nine text nodes
		expect( wrapper.find("View > Text").length ).toBe(9);
	});

	it("returns correct structure for INVALID search term", () => {	
		const wrapper = shallow(<View>{highlightTerms(CONTENT, "invalidterm", HIGHLIGHT_STYLE)}</View>);
		expect(wrapper).toMatchSnapshot();

		// There should be one text node
		expect( wrapper.find("View > Text").length ).toBe(1);
	});
});