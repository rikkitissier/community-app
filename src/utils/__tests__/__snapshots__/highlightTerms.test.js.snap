// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlightTerms utility returns correct structure for INVALID search term 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="0"
  >
    Cras mattis consectetur purus sit amet fermentum. Cum sociis natoque mattis penatibus et magnis dis parturient montes, nascetur ridiculus mus.
  </Text>
</View>
`;

exports[`highlightTerms utility returns correct structure for multiple VALID search term 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="0"
  >
    Cras
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="1"
  >
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="2"
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
     
    mattis
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="3"
  >
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="4"
  >
    consectetur purus sit amet fermentum. Cum sociis natoque
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="5"
  >
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="6"
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
     
    mattis
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="7"
  >
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="8"
  >
    penatibus et magnis dis parturient montes, nascetur ridiculus mus.
  </Text>
</View>
`;

exports[`highlightTerms utility returns correct structure for single VALID search term 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="0"
  >
    Cras mattis consectetur
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="1"
  >
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="2"
    style={
      Object {
        "backgroundColor": "red",
      }
    }
  >
     
    purus
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="3"
  >
     
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    key="4"
  >
    sit amet fermentum. Cum sociis natoque mattis penatibus et magnis dis parturient montes, nascetur ridiculus mus.
  </Text>
</View>
`;
